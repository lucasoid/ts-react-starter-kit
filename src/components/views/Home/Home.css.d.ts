export const menuItem: string;
export const dark: string;
export const light: string;
