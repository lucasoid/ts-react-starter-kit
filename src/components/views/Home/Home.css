.menuItem {
    padding: 0.5em 0;
    margin: 0;
    border-bottom: 1px solid #efefef;
}
