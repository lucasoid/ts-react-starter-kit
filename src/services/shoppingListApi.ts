export interface IList {
    id: string;
    name: string;
    owner: string;
    members?: string[];
}

export interface IListItem {
    id: string;
    listId: string;
    category: string;
    name: string;
    quantity: number;
    isActive: boolean;
}

const LATENCY: number = 1000; // to mock the latency of the API in ms

let sampleListData: IList[] = [
    {
        id: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        name: 'Weekly shopping list',
        owner: 'me@domain.com',
        members: ['you@domain.com'],
    },
    {
        id: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        name: 'Hardware supplies',
        owner: 'me@domain.com',
        members: ['him@domain.com'],
    },
    {
        id: '2a335b66-a92a-4fb0-94e7-4ec1c673937e',
        name: 'Christmas list - East coast fam',
        owner: 'you@domain.com',
        members: ['me@domain.com', 'her@domain.com'],
    },
];

let sampleListItemData: IListItem[] = [
    {
        id: '3dd08312-7cca-4467-b443-09296cb54245',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'avocados',
        quantity: 3,
        isActive: true,
    },
    {
        id: '7e81ee74-5fc1-4395-be2b-4b4616f6ab59',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'tomatoes',
        quantity: 1,
        isActive: true,
    },
    {
        id: '43529c83-a355-4099-b6f5-2179f210d4b9',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'onions',
        quantity: 1,
        isActive: true,
    },
    {
        id: '36de2c94-bfc6-41dc-91f7-6fd812107785',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'cilantro - bunch',
        quantity: 1,
        isActive: true,
    },
    {
        id: '1bf8f08a-7ddd-45b9-b401-ca8e13870382',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'spices',
        name: 'salt',
        quantity: 1,
        isActive: false,
    },
    {
        id: '293df6d2-e9d6-4f42-92f6-257543d6dde4',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'spices',
        name: 'pepper',
        quantity: 1,
        isActive: false,
    },
    {
        id: '8253fc7c-14fd-47e7-ad08-4b560cfb34ef',
        listId: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        category: 'Lumber',
        name: "2x4s - 4'",
        quantity: 8,
        isActive: true,
    },
    {
        id: '7e1b906c-f6d4-467b-9cea-7072d4cb1d43',
        listId: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        category: 'Lighting',
        name: 'LED floodlight bulbs - 2700k',
        quantity: 2,
        isActive: true,
    },
];

let listSubscribers: Function[] = [];
let listItemSubscribers: Object = {};

export const subscribeToLists = (cb: Function) => {
    if (listSubscribers.indexOf(cb) === -1) listSubscribers.push(cb);
};

export const unsubscribeToLists = (cb: Function) => {
    let index = listSubscribers.indexOf(cb);
    if (index !== -1) {
        listSubscribers.splice(index, 1);
    }
};

export const subscribeToListItems = (listId: string, cb: Function) => {
    if (listItemSubscribers[listId] === undefined) {
        listItemSubscribers[listId] = [cb];
    } else {
        listItemSubscribers[listId] = [...listItemSubscribers[listId], cb];
    }
};

export const unsubscribeToListItems = (listId: string, cb: Function) => {
    if (listItemSubscribers[listId] !== undefined) {
        let index = listItemSubscribers[listId].indexOf(cb);
        if (index !== -1) {
            listItemSubscribers[listId].splice(index, 1);
        }
    }
};

const onUpdateLists = (data: IList[]) => {
    listSubscribers.forEach(fn => fn(data));
};

const onUpdateListItems = (listId: string, data: IListItem[]) => {
    if (listItemSubscribers[listId] !== undefined) {
        listItemSubscribers[listId].forEach(fn => fn(data));
    }
};

export const fetchLists = async () => {
    const lists: IList[] = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(sampleListData);
        }, LATENCY);
    });
    onUpdateLists(lists);
    return lists;
};

export const createList = async (list: IList) => {
    const newList: IList = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(list);
        }, LATENCY);
    });
    sampleListData.push(newList);
    onUpdateLists(sampleListData);
    return list;
};

export const editList = async (list: IList) => {
    const saved: IList = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(saved);
        }, LATENCY);
    });
    const index = sampleListData.findIndex(_list => _list.id === list.id);
    sampleListData.splice(index, 1, saved);
    onUpdateLists(sampleListData);
    return saved;
};

export const fetchListItems = async (listId: string) => {
    const items: IListItem[] = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(sampleListItemData.filter(item => item.listId === listId));
        }, LATENCY);
    });
    onUpdateListItems(listId, items);
    return items;
};

export const createListItem = async (item: IListItem) => {
    const saved: IListItem = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(item);
        }, LATENCY);
    });
    sampleListItemData.push(saved);
    onUpdateListItems(item.listId, sampleListItemData);
    return item;
};

export const editListItem = async (item: IListItem) => {
    const saved: IListItem = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(item);
        }, LATENCY);
    });
    const index = sampleListItemData.findIndex(_item => _item.id === item.id);
    sampleListItemData.splice(index, 1, saved);
    onUpdateListItems(
        item.listId,
        sampleListItemData.filter(_item => _item.listId === item.listId),
    );
    return saved;
};
