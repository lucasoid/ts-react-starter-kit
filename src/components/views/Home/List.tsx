import * as React from 'react';
import { Link } from 'react-router-dom';
import { Loading } from '~components/loading';
import * as styles from './Home.css';
import { ThemeContext, Themes } from '~theme';
import { IList } from '~services/shoppingListApi';

interface IListProps {
    list: IList;
}
export const List: React.FC<IListProps> = props => {
    let { theme } = React.useContext(ThemeContext);

    return (
        <div className={`${styles.menuItem} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            {!props.list ? (
                <>
                    <Loading el={'a'} lines={1} width={'35%'} />
                    <Loading el={'aside'} lines={1} width={'25%'} />
                </>
            ) : (
                <>
                    <Link to={`/lists/${props.list.id}`} title={props.list.name}>
                        {props.list.name}
                    </Link>
                    <aside>Created by {props.list.owner}</aside>
                </>
            )}
        </div>
    );
};
