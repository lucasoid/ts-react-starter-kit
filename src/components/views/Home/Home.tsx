import * as React from 'react';
import { View } from '~components/views/View';
import { Button, ButtonTypes } from '~components/ui/Button';
import { Modal, ModalWidth } from '~components/ui/Modal';
import { ListForm } from '~components/forms/List';
import { List } from './List';
import { fetchLists, IList, subscribeToLists, unsubscribeToLists, createList } from '~services/shoppingListApi';
import { Await } from '~components/loading';

interface IHomeState {
    lists: IList[];
    creating: boolean;
    listsLoaded: boolean;
}

export class Home extends React.Component<{}, IHomeState> {
    state = {
        lists: [],
        creating: false,
        listsLoaded: false,
    };

    subscriber = lists => {
        this.setState({ lists: lists, listsLoaded: true });
    };

    componentDidMount() {
        subscribeToLists(this.subscriber);
        fetchLists();
    }

    componentWillUnmount() {
        unsubscribeToLists(this.subscriber);
    }

    openCreateDialog = () => {
        this.setState({ creating: true });
    };

    closeCreateDialog = () => {
        this.setState({ creating: false });
    };

    createList = (item: IList) => {
        // update the API, which will trigger the subscriber when it completes
        createList(item);
        // update state optimistically while we wait
        this.setState({ creating: false, lists: this.state.lists.concat(item) });
    };

    render() {
        return (
            <>
                <View>
                    <h2>My lists</h2>
                    <Await
                        loaded={this.state.listsLoaded}
                        renderFallback={() => (
                            <>
                                <List key={1} list={null} />
                                <List key={2} list={null} />
                                <List key={3} list={null} />
                            </>
                        )}
                    >
                        {this.state.lists.map(list => (
                            <List key={list.id} list={list} />
                        ))}
                        <Button
                            type={ButtonTypes.PRIMARY}
                            onClick={this.openCreateDialog}
                            styles={{ marginTop: '1em' }}
                        >
                            + New list
                        </Button>
                    </Await>
                </View>
                {/* demonstrates React.Portal. In real life I'd rather use inline editing. */}
                {this.state.creating && (
                    <Modal width={ModalWidth.MEDIUM} onDismiss={this.closeCreateDialog}>
                        <ListForm onSubmit={this.createList} onCancel={this.closeCreateDialog} />
                    </Modal>
                )}
            </>
        );
    }
}
