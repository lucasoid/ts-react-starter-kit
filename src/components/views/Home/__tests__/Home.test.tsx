import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Home } from '../Home';
// imported for mocking
import { fetchLists, IList, subscribeToLists, unsubscribeToLists, createList } from '~services/shoppingListApi';

jest.mock('~services/shoppingListApi', () => ({
    fetchLists: jest.fn(),
    subscribeToLists: jest.fn(),
    createList: jest.fn(),
    unsubscribeToLists: jest.fn(),
}));

describe('Home', () => {
    it('renders a list of lists', () => {
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb =>
            cb([
                { id: '1', name: 'My List', owner: 'me@domain.com', members: [] },
                { id: '2', name: 'My other list', owner: 'me@domain.com', members: ['you@domain.com'] },
            ]),
        );
        let wrapper = shallow(<Home />);
        expect(wrapper).toMatchSnapshot();
    });

    it('unsubscribes from the data source on unmount', () => {
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb => cb([]));
        let wrapper = shallow(<Home />);
        wrapper.unmount();
        expect(unsubscribeToLists).toBeCalled();
    });

    it('renders a loading UI', () => {
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb => null); // cb doesn't get called
        let wrapper = mount(<Home />);
        expect(wrapper).toMatchSnapshot();
    });

    it('allows for opening and closing the create dialog', () => {
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb =>
            cb([
                { id: '1', name: 'My List', owner: 'me@domain.com', members: [] },
                { id: '2', name: 'My other list', owner: 'me@domain.com', members: ['you@domain.com'] },
            ]),
        );
        let wrapper = shallow(<Home />);
        wrapper.instance().openCreateDialog();
        expect(wrapper).toMatchSnapshot();

        wrapper.instance().closeCreateDialog();
        expect(wrapper).toMatchSnapshot();
    });

    it('provides for creating a new list', () => {
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb => cb([]));
        let wrapper = shallow(<Home />);
        wrapper.instance().openCreateDialog();
        let list = { id: '3', name: 'My new list', owner: 'me@domain.com', members: [] };
        // @ts-ignore
        subscribeToLists.mockImplementationOnce(cb => cb([list]));
        wrapper.instance().createList(list);
        expect(createList).toBeCalled();
        // expect dialog to be closed and lists updated
        expect(wrapper).toMatchSnapshot();
    });
});
