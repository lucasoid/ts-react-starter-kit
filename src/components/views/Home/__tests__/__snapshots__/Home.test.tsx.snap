// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home allows for opening and closing the create dialog 1`] = `
<Fragment>
  <View>
    <h2>
      My lists
    </h2>
    <Await
      loaded={true}
      renderFallback={[Function]}
    >
      <List
        key="1"
        list={
          Object {
            "id": "1",
            "members": Array [],
            "name": "My List",
            "owner": "me@domain.com",
          }
        }
      />
      <List
        key="2"
        list={
          Object {
            "id": "2",
            "members": Array [
              "you@domain.com",
            ],
            "name": "My other list",
            "owner": "me@domain.com",
          }
        }
      />
      <Button
        onClick={[Function]}
        styles={
          Object {
            "marginTop": "1em",
          }
        }
        type={0}
      >
        + New list
      </Button>
    </Await>
  </View>
  <Modal
    onDismiss={[Function]}
    width={1}
  >
    <ListForm
      onCancel={[Function]}
      onSubmit={[Function]}
    />
  </Modal>
</Fragment>
`;

exports[`Home allows for opening and closing the create dialog 2`] = `
<Fragment>
  <View>
    <h2>
      My lists
    </h2>
    <Await
      loaded={true}
      renderFallback={[Function]}
    >
      <List
        key="1"
        list={
          Object {
            "id": "1",
            "members": Array [],
            "name": "My List",
            "owner": "me@domain.com",
          }
        }
      />
      <List
        key="2"
        list={
          Object {
            "id": "2",
            "members": Array [
              "you@domain.com",
            ],
            "name": "My other list",
            "owner": "me@domain.com",
          }
        }
      />
      <Button
        onClick={[Function]}
        styles={
          Object {
            "marginTop": "1em",
          }
        }
        type={0}
      >
        + New list
      </Button>
    </Await>
  </View>
</Fragment>
`;

exports[`Home provides for creating a new list 1`] = `
<Fragment>
  <View>
    <h2>
      My lists
    </h2>
    <Await
      loaded={true}
      renderFallback={[Function]}
    >
      <List
        key="3"
        list={
          Object {
            "id": "3",
            "members": Array [],
            "name": "My new list",
            "owner": "me@domain.com",
          }
        }
      />
      <Button
        onClick={[Function]}
        styles={
          Object {
            "marginTop": "1em",
          }
        }
        type={0}
      >
        + New list
      </Button>
    </Await>
  </View>
</Fragment>
`;

exports[`Home renders a list of lists 1`] = `
<Fragment>
  <View>
    <h2>
      My lists
    </h2>
    <Await
      loaded={true}
      renderFallback={[Function]}
    >
      <List
        key="1"
        list={
          Object {
            "id": "1",
            "members": Array [],
            "name": "My List",
            "owner": "me@domain.com",
          }
        }
      />
      <List
        key="2"
        list={
          Object {
            "id": "2",
            "members": Array [
              "you@domain.com",
            ],
            "name": "My other list",
            "owner": "me@domain.com",
          }
        }
      />
      <Button
        onClick={[Function]}
        styles={
          Object {
            "marginTop": "1em",
          }
        }
        type={0}
      >
        + New list
      </Button>
    </Await>
  </View>
</Fragment>
`;

exports[`Home renders a loading UI 1`] = `
<Home>
  <View>
    <div
      className="view light"
    >
      <ErrorBoundary>
        <h2>
          My lists
        </h2>
        <Await
          loaded={false}
          renderFallback={[Function]}
        >
          <List
            key="1"
            list={null}
          >
            <div
              className="menuItem light"
            >
              <Loading
                el="a"
                lines={1}
                width="35%"
              >
                <a
                  key="0"
                  style={
                    Object {
                      "width": "35%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </a>
              </Loading>
              <Loading
                el="aside"
                lines={1}
                width="25%"
              >
                <aside
                  key="0"
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </aside>
              </Loading>
            </div>
          </List>
          <List
            key="2"
            list={null}
          >
            <div
              className="menuItem light"
            >
              <Loading
                el="a"
                lines={1}
                width="35%"
              >
                <a
                  key="0"
                  style={
                    Object {
                      "width": "35%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </a>
              </Loading>
              <Loading
                el="aside"
                lines={1}
                width="25%"
              >
                <aside
                  key="0"
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </aside>
              </Loading>
            </div>
          </List>
          <List
            key="3"
            list={null}
          >
            <div
              className="menuItem light"
            >
              <Loading
                el="a"
                lines={1}
                width="35%"
              >
                <a
                  key="0"
                  style={
                    Object {
                      "width": "35%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </a>
              </Loading>
              <Loading
                el="aside"
                lines={1}
                width="25%"
              >
                <aside
                  key="0"
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <span
                    className="loadingText light"
                  />
                </aside>
              </Loading>
            </div>
          </List>
        </Await>
      </ErrorBoundary>
    </div>
  </View>
</Home>
`;
